src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2251 @@
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    mi->chosen_bin = i;
+  }
+
+  if (i == state->histogram_len-1) {
+    if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE &&
+        mi->histogram[i] > 0) {
+      mi->histogram[i]--;
+    }
+
+    // Infinity: Don't send a padding packet. Wait for a real packet
+    // and then see if our bins are empty or what else we should do.
+    return CIRCPAD_DELAY_INFINITE;
+  }
+
+  tor_assert(i < state->histogram_len - 1);
+
+  bin_start = circpad_histogram_bin_to_usec(mi, i);
+  bin_end = circpad_histogram_bin_to_usec(mi, i+1);
+
+  /* Truncate the high bin in case it's the infinity bin:
+   * Don't actually schedule an "infinite"-1 delay */
+  bin_end = MIN(bin_end, start_usec+state->range_sec*USEC_PER_SEC*2);
+
+  // Sample uniformly between histogram[i] to histogram[i+1]-1,
+  // but no need to sample if they are the same timeval (aka bin 0).
+  if (bin_end <= bin_start+1)
+    return bin_start;
+  else
+    return (circpad_delay_t)crypto_rand_uint64_range(bin_start, bin_end);
+}
+
+/**
+ * Sample a value from the specified probability distribution.
+ *
+ * This performs inverse transform sampling
+ * (https://en.wikipedia.org/wiki/Inverse_transform_sampling).
+ *
+ * XXX: These formulas were taken verbatim. Need a floating wizard
+ * to check them for catastropic cancellation and other issues (teor?).
+ * Also: is 32bits of double from [0.0,1.0) enough?
+ */
+static double
+circpad_distribution_sample(circpad_distribution_t dist)
+{
+  double p = 0;
+
+  switch (dist.type) {
+    case CIRCPAD_DIST_NONE:
+      return 0;
+    case CIRCPAD_DIST_UNIFORM:
+      p = crypto_rand_double();
+      // param2 is upper bound, param1 is lower
+      p *= (dist.param2 - dist.param1);
+      p += dist.param1;
+      return p;
+    case CIRCPAD_DIST_LOGISTIC:
+      while ((p = crypto_rand_double()) <= 0)
+        continue;
+      /* https://en.wikipedia.org/wiki/Logistic_distribution#Quantile_function
+       * param1 is Mu, param2 is s. */
+      return dist.param1 + dist.param2*tor_mathlog(p/(1.0-p));
+    case CIRCPAD_DIST_LOG_LOGISTIC:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Log-logistic_distribution#Quantiles
+       * param1 is Alpha, param2 is Beta */
+      return dist.param1 * pow(p/(1.0-p), 1.0/dist.param2);
+    case CIRCPAD_DIST_GEOMETRIC:
+      p = crypto_rand_double();
+      /* https://github.com/distributions-io/geometric-quantile/
+       * param1 is 'p' (success probability) */
+      return ceil(log1p(-p)/log1p(-dist.param1));
+    case CIRCPAD_DIST_WEIBULL:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Weibull_distribution \
+       *    #Cumulative_distribution_function
+       * param1 is k, param2 is Lambda */
+      return dist.param2*pow(-log1p(p), 1.0/dist.param1);
+    case CIRCPAD_DIST_PARETO:
+      p = 1.0-crypto_rand_double(); // Pareto quantile needs (0,1]
```

Comment:
Let's make crypto_rand_double() take flags for including the endpoints. (Or let's make a new function that takes those flags, and implement crypto_rand_double() using that function.)

Reply:
Hrmm.. I am kinda against a bunch of flags if this is the only case we need it for.

I could easily write up a crypto_rand_double64() though, and use that here.

Reply:
I am suggesting we add two flags: allow_zero, and allow_one.
I'll also need these flags when I rewrite the Laplace code.

Reply:
The current implementation of crypto_rand_double() tries to use whatever INT is on the platform as the underlying range for the double. I am saying I could make one that always uses uint64_t.